File: frontend/src/api/ops.ts
/**
 * Operations API client for vibe-graph-ops endpoints.
 *
 * These endpoints provide REST access to all vibe-graph operations,
 * mirroring the CLI commands.
 */

import type { ApiResponse } from "./types";

const BASE_URL = "/api/ops";

// =============================================================================
// Types
// =============================================================================

export type WorkspaceKind =
  | { type: "single_repo" }
  | { type: "multi_repo"; repo_count: number }
  | { type: "plain_directory" };

export interface WorkspaceInfo {
  name: string;
  root: string;
  kind: WorkspaceKind;
  repo_paths: string[];
}

export interface Source {
  path: string;
  relative_path: string;
  format: string;
  size: number | null;
  content: string | null;
}

export interface Repository {
  name: string;
  url: string;
  local_path: string;
  sources: Source[];
}

export interface Project {
  name: string;
  source: Record<string, unknown>;
  repositories: Repository[];
}

export interface Manifest {
  version: number;
  name: string;
  root: string;
  kind: string;
  last_sync: { secs_since_epoch: number; nanos_since_epoch: number };
  repo_count: number;
  source_count: number;
  total_size: number;
  remote: string | null;
}

// Response types
export interface SyncResponse {
  project: Project;
  workspace: WorkspaceInfo;
  path: string;
  remote: string | null;
  snapshot_created: string | null;
}

export interface GraphResponse {
  graph: {
    nodes: unknown[];
    edges: unknown[];
    metadata: Record<string, string>;
  };
  saved_path: string;
  output_path: string | null;
  from_cache: boolean;
}

export interface StatusResponse {
  workspace: WorkspaceInfo;
  store_exists: boolean;
  manifest: Manifest | null;
  snapshot_count: number;
  store_size: number;
  repositories: string[];
}

export interface LoadResponse {
  project: Project;
  manifest: Manifest;
}

export interface CleanResponse {
  cleaned: boolean;
  path: string;
}

export interface GitChangesResponse {
  changes: Array<{
    path: string;
    kind: string;
    staged: boolean;
  }>;
  repo_paths: string[];
}

export interface OpsError {
  code: string;
  message: string;
}

// =============================================================================
// API Helpers
// =============================================================================

async function opsGet<T>(endpoint: string): Promise<T> {
  const res = await fetch(`${BASE_URL}${endpoint}`);
  const json = (await res.json()) as ApiResponse<T | OpsError>;

  if (!res.ok) {
    const error = json.data as OpsError;
    throw new Error(error.message || `API error: ${res.status}`);
  }

  return json.data as T;
}

async function opsPost<T>(
  endpoint: string,
  body: Record<string, unknown>
): Promise<T> {
  const res = await fetch(`${BASE_URL}${endpoint}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  const json = (await res.json()) as ApiResponse<T | OpsError>;

  if (!res.ok) {
    const error = json.data as OpsError;
    throw new Error(error.message || `API error: ${res.status}`);
  }

  return json.data as T;
}

async function opsDelete<T>(endpoint: string): Promise<T> {
  const res = await fetch(`${BASE_URL}${endpoint}`, { method: "DELETE" });
  const json = (await res.json()) as ApiResponse<T | OpsError>;

  if (!res.ok) {
    const error = json.data as OpsError;
    throw new Error(error.message || `API error: ${res.status}`);
  }

  return json.data as T;
}

// =============================================================================
// Operations API
// =============================================================================

/**
 * Sync a codebase (local path or GitHub).
 */
export async function sync(
  source: string,
  options: {
    ignore?: string[];
    noSave?: boolean;
    snapshot?: boolean;
    useCache?: boolean;
    force?: boolean;
  } = {}
): Promise<SyncResponse> {
  return opsPost<SyncResponse>("/sync", {
    source: { type: "local", path: source },
    ignore: options.ignore || [],
    no_save: options.noSave || false,
    snapshot: options.snapshot || false,
    use_cache: options.useCache || false,
    force: options.force || false,
  });
}

/**
 * Sync using query parameters (simpler for local paths).
 */
export async function syncQuery(
  source: string,
  options: {
    ignore?: string;
    noSave?: boolean;
    useCache?: boolean;
    force?: boolean;
  } = {}
): Promise<SyncResponse> {
  const params = new URLSearchParams({ source });
  if (options.ignore) params.set("ignore", options.ignore);
  if (options.noSave) params.set("no_save", "true");
  if (options.useCache) params.set("use_cache", "true");
  if (options.force) params.set("force", "true");

  return opsGet<SyncResponse>(`/sync?${params}`);
}

/**
 * Build source code graph.
 */
export async function buildGraph(
  path: string,
  options: { force?: boolean } = {}
): Promise<GraphResponse> {
  return opsPost<GraphResponse>("/graph", {
    path,
    output: null,
    force: options.force || false,
  });
}

/**
 * Build graph using query parameters.
 */
export async function buildGraphQuery(
  path: string,
  options: { force?: boolean } = {}
): Promise<GraphResponse> {
  const params = new URLSearchParams({ path });
  if (options.force) params.set("force", "true");

  return opsGet<GraphResponse>(`/graph?${params}`);
}

/**
 * Get workspace status.
 */
export async function getStatus(
  path: string,
  options: { detailed?: boolean } = {}
): Promise<StatusResponse> {
  const params = new URLSearchParams({ path });
  if (options.detailed) params.set("detailed", "true");

  return opsGet<StatusResponse>(`/status?${params}`);
}

/**
 * Load project from .self store.
 */
export async function loadProject(path: string): Promise<LoadResponse> {
  return opsGet<LoadResponse>(`/load?path=${encodeURIComponent(path)}`);
}

/**
 * Clean .self folder.
 */
export async function clean(path: string): Promise<CleanResponse> {
  return opsDelete<CleanResponse>(`/clean?path=${encodeURIComponent(path)}`);
}

/**
 * Get git changes for workspace.
 */
export async function getGitChanges(path: string): Promise<GitChangesResponse> {
  return opsGet<GitChangesResponse>(
    `/git-changes?path=${encodeURIComponent(path)}`
  );
}

File: crates/vibe-graph-api/src/routes/ops.rs
//! Operations API endpoints.
//!
//! These endpoints provide REST access to vibe-graph operations.
//! They mirror the CLI commands but are designed for programmatic access.

use std::sync::Arc;

use axum::{
    extract::{Query, State},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use serde::{Deserialize, Serialize};
use tracing::{error, info};
use vibe_graph_ops::{
    CleanRequest, GitChangesRequest, GraphRequest, LoadRequest, OpsContext, StatusRequest,
    SyncRequest,
};

use crate::types::ApiResponse;

/// Shared state for operations endpoints.
pub struct OpsState {
    /// The operations context.
    pub ctx: OpsContext,
}

// =============================================================================
// Request/Response Types
// =============================================================================

/// Query parameters for sync endpoint.
#[derive(Debug, Deserialize)]
pub struct SyncQuery {
    /// Source to sync (path, org name, or owner/repo).
    pub source: String,
    /// Repositories to ignore (comma-separated).
    #[serde(default)]
    pub ignore: Option<String>,
    /// Whether to skip saving to .self.
    #[serde(default)]
    pub no_save: Option<bool>,
    /// Whether to use global cache.
    #[serde(default)]
    pub use_cache: Option<bool>,
    /// Whether to force fresh sync.
    #[serde(default)]
    pub force: Option<bool>,
}

/// Query parameters for graph endpoint.
#[derive(Debug, Deserialize)]
pub struct GraphQuery {
    /// Path to workspace.
    pub path: String,
    /// Force rebuild even if cached.
    #[serde(default)]
    pub force: Option<bool>,
}

/// Query parameters for status endpoint.
#[derive(Debug, Deserialize)]
pub struct StatusQuery {
    /// Path to check.
    pub path: String,
    /// Include detailed info.
    #[serde(default)]
    pub detailed: Option<bool>,
}

/// Query parameters for load endpoint.
#[derive(Debug, Deserialize)]
pub struct LoadQuery {
    /// Path to workspace.
    pub path: String,
}

/// Query parameters for clean endpoint.
#[derive(Debug, Deserialize)]
pub struct CleanQuery {
    /// Path to workspace.
    pub path: String,
}

/// Query parameters for git changes endpoint.
#[derive(Debug, Deserialize)]
pub struct GitChangesQuery {
    /// Path to workspace.
    pub path: String,
}

/// Error response for operations.
#[derive(Debug, Serialize)]
pub struct OpsErrorResponse {
    /// Error code.
    pub code: String,
    /// Error message.
    pub message: String,
}

// =============================================================================
// Handlers
// =============================================================================

/// POST /api/ops/sync - Sync a codebase.
pub async fn sync_handler(
    State(state): State<Arc<OpsState>>,
    Json(request): Json<SyncRequest>,
) -> impl IntoResponse {
    info!("Sync request: {:?}", request.source);

    match state.ctx.sync(request).await {
        Ok(response) => {
            info!(
                repos = response.repo_count(),
                files = response.file_count(),
                "Sync completed"
            );
            (StatusCode::OK, Json(ApiResponse::new(response))).into_response()
        }
        Err(e) => {
            error!("Sync failed: {}", e);
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(ApiResponse::new(OpsErrorResponse {
                    code: "SYNC_ERROR".to_string(),
                    message: e.to_string(),
                })),
            )
                .into_response()
        }
    }
}

/// GET /api/ops/sync - Sync with query parameters.
pub async fn sync_query_handler(
    State(state): State<Arc<OpsState>>,
    Query(query): Query<SyncQuery>,
) -> impl IntoResponse {
    let mut request = SyncRequest::detect(&query.source);

    if let Some(ignore) = query.ignore {
        request.ignore = ignore.split(',').map(|s| s.trim().to_string()).collect();
    }
    if query.no_save.unwrap_or(false) {
        request.no_save = true;
    }
    if query.use_cache.unwrap_or(false) {
        request.use_cache = true;
    }
    if query.force.unwrap_or(false) {
        request.force = true;
    }

    info!("Sync request (query): {:?}", request.source);

    match state.ctx.sync(request).await {
        Ok(response) => (StatusCode::OK, Json(ApiResponse::new(response))).into_response(),
        Err(e) => {
            error!("Sync failed: {}", e);
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(ApiResponse::new(OpsErrorResponse {
                    code: "SYNC_ERROR".to_string(),
                    message: e.to_string(),
                })),
            )
                .into_response()
        }
    }
}

/// POST /api/ops/graph - Build source code graph.
pub async fn graph_handler(
    State(state): State<Arc<OpsState>>,
    Json(request): Json<GraphRequest>,
) -> impl IntoResponse {
    info!("Graph request: {:?}", request.path);

    match state.ctx.graph(request).await {
        Ok(response) => {
            info!(
                nodes = response.node_count(),
                edges = response.edge_count(),
                from_cache = response.from_cache,
                "Graph built"
            );
            (StatusCode::OK, Json(ApiResponse::new(response))).into_response()
        }
        Err(e) => {
            error!("Graph build failed: {}", e);
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(ApiResponse::new(OpsErrorResponse {
                    code: "GRAPH_ERROR".to_string(),
                    message: e.to_string(),
                })),
            )
                .into_response()
        }
    }
}

/// GET /api/ops/graph - Build graph with query parameters.
pub async fn graph_query_handler(
    State(state): State<Arc<OpsState>>,
    Query(query): Query<GraphQuery>,
) -> impl IntoResponse {
    let mut request = GraphRequest::new(&query.path);
    if query.force.unwrap_or(false) {
        request = request.force();
    }

    match state.ctx.graph(request).await {
        Ok(response) => (StatusCode::OK, Json(ApiResponse::new(response))).into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(ApiResponse::new(OpsErrorResponse {
                code: "GRAPH_ERROR".to_string(),
                message: e.to_string(),
            })),
        )
            .into_response(),
    }
}

/// GET /api/ops/status - Get workspace status.
pub async fn status_handler(
    State(state): State<Arc<OpsState>>,
    Query(query): Query<StatusQuery>,
) -> impl IntoResponse {
    let mut request = StatusRequest::new(&query.path);
    if query.detailed.unwrap_or(false) {
        request = request.detailed();
    }

    match state.ctx.status(request).await {
        Ok(response) => (StatusCode::OK, Json(ApiResponse::new(response))).into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(ApiResponse::new(OpsErrorResponse {
                code: "STATUS_ERROR".to_string(),
                message: e.to_string(),
            })),
        )
            .into_response(),
    }
}

/// GET /api/ops/load - Load project from .self store.
pub async fn load_handler(
    State(state): State<Arc<OpsState>>,
    Query(query): Query<LoadQuery>,
) -> impl IntoResponse {
    let request = LoadRequest::new(&query.path);

    match state.ctx.load(request).await {
        Ok(response) => (StatusCode::OK, Json(ApiResponse::new(response))).into_response(),
        Err(e) => {
            let status = if e.to_string().contains("not found") {
                StatusCode::NOT_FOUND
            } else {
                StatusCode::INTERNAL_SERVER_ERROR
            };
            (
                status,
                Json(ApiResponse::new(OpsErrorResponse {
                    code: "LOAD_ERROR".to_string(),
                    message: e.to_string(),
                })),
            )
                .into_response()
        }
    }
}

/// DELETE /api/ops/clean - Clean .self folder.
pub async fn clean_handler(
    State(state): State<Arc<OpsState>>,
    Query(query): Query<CleanQuery>,
) -> impl IntoResponse {
    let request = CleanRequest::new(&query.path);

    match state.ctx.clean(request).await {
        Ok(response) => (StatusCode::OK, Json(ApiResponse::new(response))).into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(ApiResponse::new(OpsErrorResponse {
                code: "CLEAN_ERROR".to_string(),
                message: e.to_string(),
            })),
        )
            .into_response(),
    }
}

/// GET /api/ops/git-changes - Get git changes for workspace.
pub async fn git_changes_handler(
    State(state): State<Arc<OpsState>>,
    Query(query): Query<GitChangesQuery>,
) -> impl IntoResponse {
    let request = GitChangesRequest::new(&query.path);

    match state.ctx.git_changes(request).await {
        Ok(response) => (StatusCode::OK, Json(ApiResponse::new(response))).into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(ApiResponse::new(OpsErrorResponse {
                code: "GIT_CHANGES_ERROR".to_string(),
                message: e.to_string(),
            })),
        )
            .into_response(),
    }
}
